### PR TITLE
ENH: `__array_function__` support for `np.lib`, part 1/2

### PR DESCRIPTION
**Pull request by shoyer**
xref #12028

np.lib.arraypad through np.lib.nanfunctions

### CODE REVIEW

File: numpy/lib/arraysetops.py [hunk 1]
```diff
@@ -36,6 +37,11 @@
     ]
 
 
+def _ediff1d_dispatcher(ary, to_end=None, to_begin=None):
+    return (ary, to_end, to_begin)
```
Comment (hameerabbasi) [inline numpy/lib/arraysetops.py]: Needs a `is not None` check.

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -3294,6 +3405,12 @@ def _ureduce(a, func, **kwargs):
     return r, keepdim
 
 
+def _median_dispatcher(
+        a, axis=None, out=None, overwrite_input=None, keepdims=None):
+    return (a, out)
```
Comment (hameerabbasi) [inline numpy/lib/function_base.py]: Needs a `None` check for `out`.

File: numpy/lib/function_base.py [hunk 3]
```diff
@@ -3583,6 +3706,12 @@ def percentile(a, q, axis=None, out=None,
         a, q, axis, out, overwrite_input, interpolation, keepdims)
 
 
+def _quantile_dispatcher(a, q, axis=None, out=None, overwrite_input=None,
+                         interpolation=None, keepdims=None):
+    return (a, q, out)
```
Comment (hameerabbasi) [inline numpy/lib/function_base.py]: Needs `None` check for out.

**Review by hameerabbasi — APPROVED** [review]
Same comment here about repeated signatures, otherwise all good.

I see that you've omitted a lot of `None` checks for this PR and I realise it will never dispatch to those, but it might be nice to make it explicit. I'll leave the decision up to you though.

**Comment by shoyer** [comment]
> I see that you've omitted a lot of None checks for this PR and I realise it will never dispatch to those, but it might be nice to make it explicit.

I opted against this style mostly because it can make the dispatchers significantly longer, e.g.,
```python
def _ediff1d_dispatcher(ary, to_end=None, to_begin=None):
    return (ary, to_end, to_begin)
```
vs
```python
def _ediff1d_dispatcher(ary, to_end=None, to_begin=None):
    yield ary
    if to_end is not None:
        yield to_end
    if to_begin is not None:
        yield to_begin
```

If this actually made a difference in speed I would go for it, but I suspect they will actually be quite similar in performance.

**Comment by charris** [comment]
Thanks Stephan.

**Pull request merged**